proposals/0000-wire-protocol.md

```diff
@@ -0,0 +1,480 @@
+
+The `payload` may be any message content, as needed by the extension.
+
+```
+<varint user-type><payload>
+```
+
+
+## Run-Length Encoded Bitfield
+[run-length-encoded-bitfield]: #run-length-encoded-bitfield
+
+The Run-Length Encoded (RLE) bitfield is a series of compressed and uncompressed bit sequences.All sequences start with a header that is a varint. If the last bit is set in the varint (it is an odd number) then a header represents a compressed bit sequence. If the last bit is not set then a header represents an non compressed sequence.
+
+```
+compressed-sequence = varint(byte-length-of-sequence << 2 | bit << 1 | 1)
+uncompressed-sequence = varint(byte-length-of-bitfield << 1 | 0) + bitfield
+```
+
+
+## Block Tree Digest
+[block-tree-digest]: #block-tree-digest
+
+As described in DEP-0002 (Hypercore), a peer should be able to verify both the integrity of received data (aka, was there corruption somewhere along the way, detected via hash) and the authenticity (aka, is this the data from the original writer, detected via signature). Hypercore transmits the hash for every data block, but only signatures for the root hashes of Merkle trees, not individual block hashes, which means a peer may need additional hashes (for data blocks they do not have a copy of) if they want to verify the signatures of individual blocks.
+
+Redundantly transmitting all such hashes on every request would be inefficient if the receiver already had some hashes, so requesting peers can specify which hashes they need in the `nodes` field of the `Request` message. Instead of sending node indexes, the peer can send a compact bitfield, indicating for each "uncle" and "parent" whether the hash should be transmitted.
+
+Consider the following tree:
+
+```
+0
+  1
+2
+    3
+4
+  5
+6
+```
+
+If the receiving peers wants to fetch, and verify, block 6, it needs to communicate whether it already has the uncle hashes (4, 1) and the parent hashes (3). This information can be compressed into a small bit vector with the following scheme:
+
+ - the least-significant bit indicates whether the most-significant bit is a "parent" (if '1') or an "uncle" (if '0')
+ - all other bits, in order from least- to most-significant, indicate whether the corresponding "uncle" hash *does* need to be transmitted (if bit '0') or *does not* (if bit '1')
+
+An an example, suppose we want to fetch block 6 from a remote peer, and we already have the sparse node metadata (hashes) for blocks 3 and 4, but not 1. In other words:
+
+ - 4, an uncle, we already have the hash
+ - 1, next uncle, we don't have hash
+ - 3, a parent (and root hash), we do have hash
+ - root signature (of hash 3) will be sent with request
+
+Our `Request` should include the digest bits:
+
+```
+101(1) <-- indicates that the most-significant bit is a parent, not an uncle
+10(1)1 <-- do not send hash for the first uncle, 4
+1(0)11 <-- do send hash for the next uncle, 1
+(1)000 <-- do not send hash for next parent, 3
```

Comment:
Why is this not `(1)011` ?

Reply:
General remark, wouldn't this be better readable like:

```
1011 
┃┃┃┗━ indicates that the most-significant bit is a parent, not an uncle
┃┃┗━━ do not send hash for the first uncle, 4
┃┗━━━ do send hash for the next uncle, 1
┗━━━━ do not send hash for next parent, 3
```

Reply:
Good catch, and that formatting is much nicer, thanks!